lib/llm/src/mocker/evictor.rs

```diff
@@ -13,100 +13,133 @@
     }
 
     /// Get an iterator over the keys in the evictor
-    pub fn keys(&self) -> std::collections::hash_map::Keys<'_, T, f64> {
+    pub fn keys(&self) -> std::collections::hash_map::Keys<'_, T, i64> {
         self.free_table.keys()
     }
 
-    /// Insert or update an object in the evictor with current timestamp
+    /// Private helper method to update the data structures with object and counter
+    fn _update(&mut self, object: T, counter: i64) {
+        self.free_table.insert(object.clone(), counter);
+        self.priority_queue.insert(PriorityItem {
+            item: object,
+            counter,
+        });
+    }
+
+    /// Insert or update an object in the evictor with positive counter
     pub fn insert(&mut self, object: T) {
-        let timestamp = self.current_timestamp();
-        self._insert(object, timestamp);
+        // Remove old entry if it exists
+        if let Some(&old_counter) = self.free_table.get(&object) {
+            self.priority_queue.remove(&PriorityItem {
+                item: object.clone(),
+                counter: old_counter,
+            });
+        }
+
+        // Increment positive counter and insert
+        self.positive_counter += 1;
+        let counter = self.positive_counter;
+
+        self._update(object, counter);
+    }
+
+    /// Push an object to the front with negative counter (highest priority for eviction)
+    pub fn push_front(&mut self, object: T) {
+        // Remove old entry if it exists
+        if let Some(&old_counter) = self.free_table.get(&object) {
+            self.priority_queue.remove(&PriorityItem {
+                item: object.clone(),
+                counter: old_counter,
+            });
+        }
+
+        // Decrement negative counter and insert
+        self.negative_counter -= 1;
+        let counter = self.negative_counter;
+
+        self._update(object, counter);
     }
 
     /// Check if the evictor contains the given object
     pub fn contains(&self, object: &T) -> bool {
         self.free_table.contains_key(object)
     }
 
-    /// Evict an object based on LRU policy
+    /// Evict an object based on LRU policy (lowest counter value)
     /// Returns the evicted object or None if no objects are available
     pub fn evict(&mut self) -> Option<T> {
-        if self.free_table.is_empty() {
-            return None;
-        }
-
-        while let Some((object, last_accessed)) = self.priority_queue.pop_front() {
-            let Some(&current_last_accessed) = self.free_table.get(&object) else {
-                continue; // entry is already removed
-            };
-
-            if current_last_accessed == last_accessed {
-                self.free_table.remove(&object);
-                return Some(object);
-            } // otherwise entry is stale
+        if let Some(item) = self.priority_queue.pop_first() {
+            self.free_table.remove(&item.item);
+            Some(item.item)
+        } else {
+            None
```

Comment:
Optional, but I think you could re-write this without the if statement, using `pop_first().map(|item| .. no spoilers` :-)